iptables/src/utils.cpp

```diff
@@ -0,0 +1,309 @@
+
+#include "utils.h"
+
+using namespace osquery;
+
+namespace trailofbits {
+static const std::string kIptablesSave = "/sbin/iptables-save";
+static const std::string kIp6tablesSave = "/sbin/ip6tables-save";
+
+static const std::string kIptablesNamesFile = "/proc/net/ip_tables_names";
+static const std::string kIp6tablesNamesFile = "/proc/net/ip6_tables_names";
+
+static const std::string kHexMap = "0123456789ABCDEF";
+
+static const int kMaskHighBits = 4;
+static const int kMaskLowBits = 15;
+
+static TableList getTableNames(const std::string& filename) {
+  std::string content;
+
+  auto s = readFile(filename, content);
+
+  if (!s.ok()) {
+    TLOG << "Error reading: " << filename << ": " << s.toString();
+    return {};
+  }
+
+  auto results = SplitString(content, '\n');
+
+  for (auto& result : results) {
+    boost::algorithm::trim(result);
+  }
+
+  return results;
+}
+
+/* NOTE(ww): This function takes either iptables-save or ip6tables-save.
+ * It would be nice to use iptables-xml and ip6tables-xml, but ip6tables-xml
+ * doesn't exist yet and iptables-xml frequently hangs (during experimentation).
+ */
+Status genMatchMap(const std::string& ipt_cmd, MatchMap& match_map) {
+  ProcessOutput output;
+
+  if (!ExecuteProcess(output, ipt_cmd, {}) || output.exit_code != 0) {
+    return Status(1, "couldn't exec " + ipt_cmd);
+  }
+
+  if (output.std_output.empty()) {
+    return Status(1, "no output from command");
+  }
+
+  std::string table;
+  std::string chain;
+  for (const auto& line : SplitString(output.std_output, '\n')) {
+    std::string match;
+    std::string target;
+    std::string target_options;
+
+    // If the line is empty or a comment, skip it.
+    if (line.empty() || line.at(0) == '#') {
+      continue;
+    }
+
+    // If the line is a filter name, record it and prep our toplevel map with
+    // it.
+    if (line.at(0) == '*') {
+      table = line.substr(1);
+      match_map[table];
+      continue;
+    }
+
+    // If the line is a chain name, record it and prep our chain map with it.
+    if (line.at(0) == ':') {
+      auto stop = line.find(" ");
+      chain = line.substr(1, stop - 1);
+      match_map[table][chain];
+      continue;
+    }
+
+    // If the line is a rule, look for match entries, targets,
+    // and target options within it.
+    if (line.find("-A") == 0) {
+      auto start = line.find(" ");
+      auto stop = line.find(" ", start + 1);
+
+      chain = line.substr(start + 1, stop - start - 1);
+
+      // Matches begin with an -m.
+      start = line.find(" -m ");
+
+      // Targets begin with a -j or a -g.
+      stop = line.rfind(" -j ");
+      if (stop == std::string::npos) {
+        stop = line.rfind(" -g ");
+      }
+
+      // Match extraction.
+      if (start == std::string::npos) {
+        match = "";
+      } else {
+        if (stop != std::string::npos && stop < start) {
+          TLOG << "Oddity: -j or -g before -m: " << line;
+          match = "";
```

Comment:
here we set match to "" even though `-m` existed -- since `-m` existed shouldn't we try to keep parsing? or do we intentionally not try to parse since it's kind of a malformed line we were not expecting?

Reply:
The latter -- `iptables-save` should never emit a `-j` or `-g` target rule before match extensions, so finding one means that something's gone really wrong.